include/ada/url_aggregator-inl.h

```diff
@@ -239,11 +240,22 @@ inline void url_aggregator::update_base_search(
     }
 
     buffer.insert(components.search_start, "?");
-
-    std::string encoded =
-        unicode::percent_encode(input, query_percent_encode_set);
-    buffer.insert(components.search_start + 1, encoded);
-    components.hash_start += uint32_t(encoded.size() + 1);  // Do not forget `?`
+    size_t idx =
+        ada::unicode::percent_encode_index(input, query_percent_encode_set);
+    if (idx == input.size()) {
+      buffer.insert(components.search_start + 1, input);
+      components.hash_start += uint32_t(input.size() + 1);  // Do not forget `?`
+    } else {
+      buffer.insert(components.search_start + 1, input, 0, idx);
+      input.remove_prefix(idx);
+      // We only create a temporary string if we need percent encoding and
+      // we attempt to create as small a temporary string as we can.
+      std::string encoded =
+          ada::unicode::percent_encode(input, query_percent_encode_set);
```

Comment:
In the future we can just pass the string and the index to this function and directly insert to the string with an internal counter to keep track the inserted char count to avoid this string creation.